draft-specifications/Model.md

```diff
@@ -0,0 +1,324 @@
+> https://github.com/ocapn/ocapn/issues/47
+>
+> Agoric previously held the position that all valid strings that JavaScript
+> can parse with `JSON.parse` should round-trip.
+> https://github.com/ocapn/ocapn/issues/5#issuecomment-1550370096,
+> https://github.com/ocapn/ocapn/issues/5#issuecomment-1550481111
+>
+> Spritely has consistently preferred a representation that can be captured
+> with UTF-8, which cannot transit lone/unpaired surrogates.
+> (October 10, 2023)
+
+## ByteArray
+
+An array of bytes.
+
+> - **Guile**: `#vu8()`
+> - **JavaScript**: `new ArrayBuffer()`
+> - **Python**: `b''`
+>
+> Tracking: https://github.com/ocapn/ocapn/issues/48
+>
+> At the [November 14, 2023](https://github.com/ocapn/ocapn/issues/94) meeting,
+> we agreed to settle on the prefix "byte" and agreed to decide at the next
+> meeting based on a [reactji
+> poll](https://github.com/ocapn/ocapn/issues/48#issuecomment-1811097196).
+>
+> [Jan 2024 meeting
+> notes](https://cryptpad.fr/code/#/2/code/edit/Ng7kXl9rCF5qnTQX60UCpDHE/)
+> record that we agreed on ByteArray because it was the winner of the poll, and
+> we had already agreed to resolve this specific issue by poll among these
+> three choices.
+
+## Selector
+
+A string of Unicode code points, differentiated from a proper string by its
+type.
+Selectors cannot contain lone/unpaired surrogates.
+
+> - **Guile**: symbols `'name`
+> - **JavaScript**:
+>   ```js
+>   {
+>     [Symbol.for('passStyle'): 'selector',
+>     [Symbol.toStringTag]: 'name',
+>   }
+>   ```
+> - **Python**: `Selector('name')` where `Selector` is imported from `ocapn`.
+>
+> Tracking: https://github.com/ocapn/ocapn/issues/46
+>
+> OCapN uses the name selector to avoid the implication that they will
+> correspond to a language's symbol type in all languages that have a symbol
+> type.
+> Selectors may correspond to symbols in languages where a symbol is eligible
+> for garbage collection when there are no extant references.
+> At this time, JavaScript cannot safely use registered symbols like
+> `Symbol.for('name')` for OCapN selectors, because registered symbols are
+> interned without possibility of eventual garbage collection.
+>
+> OCapN supports one operator for delivering both function application and
+> method invocation.
+> By convention, method invocation is equivalent to function application, where
+> the first argument is a selector followed by the remaining arguments.
+
+# Container
+
+## List (JSON)
+
+A list of any quantity of values.
+
+> - **Guile**: `'()`
+> - **JavaScript** and **JSON**: `[]`
+> - **Python**: `()` (We have not discussed whether to use tuple or list. Tuple
+>   is less mutable, though enforcing immutability in Python is not likely to be
+>   a goal the way it is in hardened JavaScript.)
+>
+> Lists participate in the JSON subset of OCapN.
+>
+> We achieved consensus to name the type "List" at the [November 14, 2023](https://github.com/ocapn/ocapn/issues/94) meeting.
+
+## Struct (JSON)
```

Comment:
The important point is that the keys must be strings, and “dictionary” does capture that distinction in juxtaposition with “map”. Agoric uses the term “record”, but I believe that was disqualified on the grounds that it is too close to a term present in the Syrup concrete model (but absent in the data model!). I am sure Agoric (@erights) would welcome the term “record” if we can move the Syrup word out of the way (say “tuple”).

Reply:
If you ask a C person, they will tell you that a struct is an association between a name (constrained to valid C identifiers) and a value (at an offset in a compact memory representation implied by their order of appearance and padding necessary to satisfy the underlying architecture). Our interpretation is consistent with the germane portion of that.

If you ask a C++ person, a struct is simply a class where all of the fields and methods are implicitly public. As in most cases, the C++ person is wrong.

A Protobuf person will tell you they prefer the term “message”. Messages can contain messages, and in fact, you can’t express the difference between presence of a zero value and absence of a field without a so-called “message”. Their opinion can be dismissed because they had an opportunity to pick one of “protocol” or “interface”, and instead did both (Protobuf IDL) and added  “service”, to refer to essentially the same thing. Protobuf people have strange preferences.

If we choose to reöpen this aging wound, I dislike but am open to “dictionary”. I like “record”. I think “struct” is fine.

Reply:
Having discussed this here and with @erights out-of-band, I am continuing this PR with “struct” since we have tentative consensus on this and not the alternatives. Mark and I object to “dictionary” since our model for dictionaries in JavaScript/TypeScript is that a dictionary is a specialization on maps, where the keys are specifically strings, and values are a homogeneous type, whereas structs are modeled as objects with known and heterogeneous (KeyT to ValueT) pairs.

```ts
type Dict<T> = Map<string, T>
```

Mark and I would not object to “record” but expect Spritely would on grounds that the term is reserved for a different concept in the concrete Syrup protocol.

Reply:
At the December 10 meeting, we decided to proceed with verbiage that the name “struct” is tentative. https://github.com/ocapn/ocapn/issues/144